alternator/expression_factory.cc

```diff
@@ -0,0 +1,188 @@
+
+    // Lookup by string_view, move key to back, and return value
+    std::optional<std::reference_wrapper<Value>> find(std::string_view key) {
+        auto it = map.find(key);
+        if(it == map.end()) {
+            return std::nullopt;
+        }
+        touch(it);
+        return std::ref(it->second);
+    }
+
+    // Removes the least recently used item from the cache
+    // Returns true if an item was removed, false if the cache was empty
+    bool pop() {
+        if (!lru_list.empty()) {
+            auto it = map.find(lru_list.front());
+            lru_list.pop_front();
+            if(it != map.end()) { // This should be always true
+                map.erase(it);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    // Insert or update the value for the given key.
+    void push(std::string_view key, Value&& value) {
+        auto ret = map.insert_or_assign(key, std::move(value));
+        (ret.second ? touch<true>(ret.first) : touch<false>(ret.first));
+    }
+};
+}
+
+// Actual parsing functions are defined in alternator/expressions.cc
+parsed::update_expression parse_update_expression(std::string_view query);
+std::vector<parsed::path> parse_projection_expression(std::string_view query);
+parsed::condition_expression parse_condition_expression(std::string_view query, const char* caller);
+
+struct expression_factory_impl {
+    expression_factory::config _cfg;
+    size_t _max_cache_entries;
+    stats& _stats;
+
+    using cached_expressions_types = std::variant<
+        parsed::update_expression,
+        parsed::condition_expression,
+        std::vector<parsed::path>
+    >;
+    lru_string_map<cached_expressions_types> _cached_entries;
+    utils::observable<uint32_t>::observer _max_cache_entries_observer;
+
+    expression_factory_impl(expression_factory::config cfg, stats& stats);
+
+    template <typename Func, typename... Args>
+    using ParseResult = std::invoke_result_t<Func, std::string_view, Args...>;
+
+    template <stats::expression_types StatsType, typename Func, typename... Args>
+    ParseResult<Func, Args...> get_or_create(std::string_view query, Func&& parse_func, Args&&... other_args) {
+        if (_cfg.max_cache_entries == 0) {
+            return parse_func(query, std::forward<Args>(other_args)...);
+        }
+        auto value = _cached_entries.find(query);
+        if (value) {
+            cache_logger.trace("Cache hit for query: {}", query);
+            _stats.expression_cache.requests[StatsType].hits++;
+            try {
+                return std::get<ParseResult<Func, Args...>>(value->get());
+            } catch (const std::bad_variant_access&) {
+                // Although nothing here prevents same query to be parsed into different types,
+                // in practice valid queries are different enough to not collide.
+                cache_logger.error("Cache hit for '{}' but type mismatch - it should never happen.", query);
```

Comment:
Avi's proposal above will solve this. I don't like this error log and the text that "should never happen" - a mischievous user can easily cause this by passing the same string to two types of expressions - e.g., pass "SET a=3" as an UpdateExpression, and then give "SET a=3" as a ConditionExpression and instead of getting the expected error, you'll get a scary error log (I hope you'll also get the parse error later, right?)